src/LabLog/Services/SchoolService.cs

```diff
@@ -0,0 +1,97 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using LabLog.Models;
+using Microsoft.EntityFrameworkCore;
+using LabLog.Domain.Events;
+
+
+namespace LabLog.Services
+{
+    public class SchoolService
+    {
+        private EventModelContext _db;
+        private string _user;
+        private SchoolEventHandler _school;
+
+        public SchoolService(EventModelContext db, string user)
+        {
+            _db = db;
+            _user = user;
+            _school = new SchoolEventHandler(_user, db);
+        }
+
+        public List<SchoolModel> RebuildReadModel()
+        {
+            DeleteReadModelFromDatabase();
+            var schoolCreateEvents = _db.LabEvents.Where(o => (o.EventType == SchoolCreatedEvent.EventTypeString));
+
+            foreach (ILabEvent e in schoolCreateEvents)
+            {
+                SchoolModel school = GetSchool(e.SchoolId);
+                _db.Add(school);
+                _db.SaveChanges();
+            }
+            List<SchoolModel> schools = _db.Schools.ToList();
+            return schools;
+        }
+
+        private void DeleteReadModelFromDatabase()
+        {
+            _db.Database.ExecuteSqlCommand("DELETE FROM ComputerModel;");
+            _db.Database.ExecuteSqlCommand("DELETE FROM RoomModel;");
+            _db.Database.ExecuteSqlCommand("DELETE FROM Schools;");
+        }
+
+        public void CreateSchool(SchoolModel school)
+        {
+                Domain.Entities.School.Create(school.Name, e =>
+                {
+                    SchoolModel _school = new SchoolModel();
+                    e.EventAuthor = _user;
+                    _db.Add(e);
+                    _school.ApplySchoolCreatedEvent(e);
+                    _db.Add(_school);
+                    _db.SaveChanges();
+                });
+        }
+
+        public SchoolModel GetSchool (Guid schoolId)
+        {
+            return (SchoolModel.GetSchoolFromDb(_db, schoolId));
+        }
+
+        public void GetSchoolRooms(SchoolModel school)
+        {
+            _db.Entry(school).Collection(c => c.Rooms).Load();
+        }
+
+        public RoomModel GetRoom(SchoolModel school, string roomName)
+        {
+            GetSchoolRooms(school);
+            return (school.GetRoom(roomName));
+        }
+
+        public void GetRoomComputers(RoomModel room)
+        {
+            _db.Entry(room).Collection(c => c.Computers).Load();
```

Comment:
I don't like this. It's ugly hackiness to deal with the lack of lazy loading in EF 7 [https://github.com/aspnet/EntityFrameworkCore/issues/3312](url)
Can you think of a better way to handle this? I don't want to eager load whole collections in advance just because I want a single record, but also don't want to have to keep re-querying the database.

Reply:
This is what .Include() is for, to include in the db query joins to the dependent tables

Reply:
Yeah, but if I want a single room, I don't want the database to be queried for EVERY room. And having already retrieved a school without all the rooms, re-querying for the same school just to be able to put in the .include seems silly too.

Reply:
nah, .Include() just tells it to load the objects that match the query.
EF uses a concept called Delayed Execution. The db query is only run when you call an ending method, like SingleOrDefault(), ToList(), First() etc.
So it'll build a query, restricting the output set to whatever "where" constraints you have, and then join into the rooms that match